lib/src/main/java/org/esbtools/eventhandler/PromiseOfPromise.java

```diff
@@ -0,0 +1,69 @@
+/*
+ *  Copyright 2016 esbtools Contributors and/or its affiliates.
+ *
+ *  This file is part of esbtools.
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.esbtools.eventhandler;
+
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+public class PromiseOfPromise<U> implements Promise<U> {
```

Comment:
> If something else resolves the Future, then if you only have a Future, you will forever block on .get(). It requires you to introduce another concept, and this breaks the encapsulation of `process()`.

This is perhaps a reason to consider changing how we convey the interface of inbound. i.e. a consequence of having the interface use `Future` (vs. my suggestion of looking at having the interface be around a collection of messages).